[Add] NanoVLM Blog Post (#2865)
huggingface/blog: nanovlm.md

---
title: "nanoVLM: The simplest repository to train your VLM in pure PyTorch" 
thumbnail: /blog/assets/nanovlm/thumbnail.png
authors:
- user: ariG23498
- user: lusxvr
- user: andito
- user: sergiopaniego
- user: merve
- user: pcuenq
- user: reach-vb
---

# nanoVLM: The simplest repository to train your VLM in pure PyTorch

Inspired by [Andrej Karpathy](https://karpathy.ai/)’s [nanoGPT](https://github.com/karpathy/nanoGPT),
Hugging Face offers a similar project for the vision domain: [**nanoVLM**](https://github.com/huggingface/nanoVLM).
This repository is designed to be the simplest way to get started with training your own
Vision Language Model (VLM) using pure PyTorch.

At its heart, nanoVLM is a **toolkit** that helps you build and train a model that can understand both
images and text, and then generate text based on that. The beauty of nanoVLM lies in its *simplicity*.
The entire codebase is intentionally kept *minimal* and *readable*, making it perfect for beginners or
anyone who wants to peek under the hood of VLMs without getting overwhelmed.

> 

Author: I feel like this should make it a little redundant. I have added the note on being lightweight in the previous para.

Could you review the changes and see how that works? If you still want me to rephrase the current para, I would happily do it.

- [Architecture](#architecture)
- [Train your own VLM](#train-your-own-vlm)
- [Run inference on a pre-trained model](#run-inference-on-a-pre-trained-model)
- [Conclusion](#conclusion)
- [References](#references)

## TL;DR

You can start training a Vision Language Model using our nanoVLM toolkit by following these steps:

```bash
# Clone the repo
git clone https://github.com/huggingface/transformers.git

# Execute the training script
python train.py
```

Here is a [Colab notebook](https://colab.research.google.com/github/huggingface/nanoVLM/blob/main/nanoVLM.ipynb)
that will help you get started with no local setup required!

## What is a Vision Language Model?

As the name suggests, a Vision Language Model (VLM) is a multi-modal model that processes two
modalities: vision and text.These models typically take images and/or text as input and generate text as output.

Generating text (output) conditioned on the understanding of images and texts (inputs) is a powerful paradigm.
It enables a wide range of applications, from image captioning and object detection to answering
questions about visual content (as shown in the table below). One thing to note is that nanoVLM

> focuses only on Visual Question Answering as the training objective.

Author: I agree with your PoV, but I had a lot of feedback on this particular para being confusing, hence I wanted to be explicit.

| Image (input) | Query (Text as input) | Generation | Task |
| :---- | :---- | :---- | :---- |
| ![an image of a cat](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/nanovlm/cat.jpg) | Caption the image | Two cats lying down on a bed with remotes near them | Captioning |
|  | Detect the objects in the image | `<locxx><locxx><locxx><locxx>` | Object Detection |
|  | Segment the objects in the image | `<segxx><segxx><segxx>` | Semantic Segmentation |
|  | How many cats are in the image? | 2 | Visual Question Answering |

> [!TIP]  
> If you are interested in learning more about VLMs, we strongly recommend reading our latest

> blog on the topic: [Vision Language Models (Better, Faster, Stronger)](https://huggingface.co/blog/vlms-2025)

Author: Great catch!

├── data
│   ├── collators.py
│   ├── datasets.py
│   └── processors.py
├── generate.py
├── models
│   ├── config.py
│   ├── language_model.py
│   ├── modality_projector.py
│   ├── README.md
│   ├── utils.py
│   ├── vision_language_model.py
│   └── vision_transformer.py
└── train.py
```

## Architecture

```bash
.
├── data
│   └── ...
├── models      # 👈 You are here
│   └── ...
└── train.py     
```

We model nanoVLM after two well known and widely used architectures. Our vision backbone
(`models/vision_transformer.py`) is the standard vision transformer, more specifically Google’s
[SigLIP](https://huggingface.co/docs/transformers/en/model_doc/siglip) vision encoder. Our language
backbone follows the [Llama 3](https://huggingface.co/docs/transformers/en/model_doc/llama3) architecture.

The vision and text modalities are *aligned* using a Modality Projection module. This module takes the
image embeddings produced by the vision backbone as input, and transforms them into embeddings
compatible with the text embeddings from the embedding layer of the language model. These embeddings
are then concatenated and fed into the language decoder. The Modality Projection module consists of a

> pixel shuffle operation followed by a linear layer.

Author: All thanks to @sergiopaniego!

| ![diagram of the model architecture](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/nanovlm/architecture.png) |
| :--: |

> | The architecture of the model (Source: Authors) |

Author: For sure. I will add the paper in the references section.

Now that we are familiar with the architecture, let's shift gears and talk about how to train your own Vision Language Model using `train.py`.

```bash
.
├── data
│   └── ...
├── models
│   └── ...
└── train.py     # 👈 You are here
```

You can kick off training with:

```bash
python train.py
```

This script is your one-stop shop for the entire training pipeline, including:

* Dataset loading and preprocessing  
* Model initialization  
* Optimization and logging

**Configuration**

Before anything else, the script loads two configuration classes from `models/config.py`:

* `TrainConfig`: Configuration parameters useful for training, like learning rates, checkpoint paths, etc.  
* `VLMConfig`: The configuration parameters used to initialize the VLM, like hidden dimensions, number of attention heads, etc.

**Data Loading**

At the heart of the data pipeline is the `get_dataloaders` function. It:

* Loads datasets via Hugging Face’s `load_dataset` API.  
* Combines and shuffles multiple datasets (if provided).  
* Applies a train/val split via indexing.  
* Wraps them in custom datasets (`VQADataset`, `MMStarDataset`) and collators (`VQACollator`, `MMStarCollator`).

> [!TIP]  

> > A helpful flag here is `data_cutoff_idx`, useful for debugging on small subsets.

Author: Would you like me to remove this? 😆

> > A helpful flag here is `data_cutoff_idx`, useful for debugging on small subsets.

Reviewer: no no, it's great, I wasn't so hands on with nanovlm and really didn't know what this was for

This part is fairly standard but thoughtfully structured:

* Mixed precision is used with `torch.autocast` to improve performance.  
* A cosine learning rate schedule with linear warmup is implemented via `get_lr`.  
* Token throughput (tokens/sec) is logged per batch for performance monitoring.

Every 250 steps (configurable), the model is evaluated on the validation and `MMStar` test datasets. If accuracy improves, the model is checkpointed.

**Logging & Monitoring**

If `log_wandb` is enabled, training stats like `batch_loss`,  `val_loss`, `accuracy`, and `tokens_per_second`
are logged to Weights & Biases for real-time tracking.

Runs are auto-named using metadata like sample size, batch size, epoch count, learning rates, and the date,
all handled by the helper `get_run_name`.

**Push to Hub**

The script doesn’t automatically push to the Hugging Face Hub, but since models are saved via:

```python
model.save_pretrained(path)
```

You can easily push them using:

```python
model.push_to_hub("hub/id")
```

## Run inference on a pre-trained model

Using nanoVLM as the toolkit, we have trained a [model and published it to Hub](https://huggingface.co/lusxvr/nanoVLM-222M).

> We have used the `SigLIP-B/16-224-85M` and `HuggingFaceTB/SmolLM2-135M` as backbones. The model was

Author: I got this from the README of nanoVLM. I will use the checkpoint shared by you then.

You can use this script on your own images and prompts like so:

```bash
python generate.py --image path/to/image.png --prompt "You prompt here"
```

If you want to visualize the heart of the script, it is just these lines:

```python
model = VisionLanguageModel.from_pretrained(source).to(device)
model.eval()

tokenizer = get_tokenizer(model.cfg.lm_tokenizer)
image_processor = get_image_processor(model.cfg.vit_img_size)

template = f"Question: {args.prompt} Answer:"
encoded = tokenizer.batch_encode_plus([template], return_tensors="pt")
tokens = encoded["input_ids"].to(device)

img = Image.open(args.image).convert("RGB")
img_t = image_processor(img).unsqueeze(0).to(device)

print("\nInput:\n ", args.prompt, "\n\nOutputs:")
for i in range(args.generations):
    gen = model.generate(tokens, img_t, max_new_tokens=args.max_new_tokens)
    out = tokenizer.batch_decode(gen, skip_special_tokens=True)[0]
    print(f"  >> Generation {i+1}: {out}")
```

We create the model and set it to `eval`. Initialize the tokenizer, which tokenizes the text prompt,
and the image processor, which  is used to process the images. The next step is to process the inputs

> and run `model.generate` to generate the output text. Finally, decode the output using  `batch_decode`.

Reviewer: Should we show the results and the image we use? Like show the cat picture and some generations? Maybe we could have a couple of pictures and generations

> [!TIP]  
> If you want to run inference on the trained model in a UI interface, [here](https://huggingface.co/spaces/ariG23498/nanovlm) is the Hugging Face Space for you to interact with the model. 

## Conclusion

In this blog post, we walked through what VLMs are, explored the architecture choices that power nanoVLM, and unpacked the training and inference workflows in detail.

By keeping the codebase lightweight and readable, nanoVLM aims to serve as both a learning tool and a foundation you can build upon. Whether you’re looking to understand how multi-modal inputs are aligned, or you want to train a VLM on your own dataset, this repository gives you a head start.

If you try it out, build on top of it, or just have questions we’d love to hear from you. Happy tinkering!

## References

1. [GitHub - huggingface/nanoVLM: The simplest, fastest repository for training/finetuning small-sized VLMs.](https://github.com/huggingface/nanoVLM)  
2. [A Dive into Vision-Language Models](https://huggingface.co/blog/vision_language_pretraining)  

> 3. [Vision Language Models Explained](https://huggingface.co/blog/vlms)  

Reviewer: Maybe add the SmolVLm paper here?

- [References](#references)

## TL;DR

You can start training a Vision Language Model using our nanoVLM toolkit by following these steps:

```bash
# Clone the repo
git clone https://github.com/huggingface/transformers.git

# Execute the training script
python train.py
```

Here is a [Colab notebook](https://colab.research.google.com/github/huggingface/nanoVLM/blob/main/nanoVLM.ipynb)
that will help you launch a trainig run with no local setup required!

## What is a Vision Language Model?

As the name suggests, a Vision Language Model (VLM) is a multi-modal model that processes two
modalities: vision and text. These models typically take images and/or text as input and generate text as output.

Generating text (output) conditioned on the understanding of images and texts (inputs) is a powerful paradigm.
It enables a wide range of applications, from image captioning and object detection to answering
questions about visual content (as shown in the table below). One thing to note is that nanoVLM
focuses only on Visual Question Answering as the training objective.

| Image (input) | Query (Text as input) | Generation | Task |
| :---- | :---- | :---- | :---- |

> | ![an image of a cat](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/nanovlm/cat.jpg) | Caption the image | Two cats lying down on a bed with remotes near them | Captioning |

Reviewer 2: Is it possible to have the image be over all the lines of the table? I.e. merge the first column into one row?

Reviewer 3: Yes (using HTML). I've added it on a suggestion

> | Image (input) | Query (Text as input) | Generation | Task |
> | :---- | :---- | :---- | :---- |
> | ![an image of a cat](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/nanovlm/cat.jpg) | Caption the image | Two cats lying down on a bed with remotes near them | Captioning |
> |  | Detect the objects in the image | `<locxx><locxx><locxx><locxx>` | Object Detection |
> |  | Segment the objects in the image | `<segxx><segxx><segxx>` | Semantic Segmentation |
> |  | How many cats are in the image? | 2 | Visual Question Answering |

Reviewer 3: ```suggestion
<table>
  <tr>
    <td rowspan="4"><img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/nanovlm/cat.jpg" alt="an image of a cat" width="200"/></td>
    <td>Caption the image</td>
    <td>Two cats lying down on a bed with remotes near them</td>
    <td>Captioning</td>
  </tr>
  <tr>
    <td>Detect the objects in the image</td>
    <td><code>&lt;locxx&gt;&lt;locxx&gt;&lt;locxx&gt;&lt;locxx&gt;</code></td>
    <td>Object Detection</td>
  </tr>
  <tr>
    <td>Segment the objects in the image</td>
    <td><code>&lt;segxx&gt;&lt;segxx&gt;&lt;segxx&gt;</code></td>
    <td>Semantic Segmentation</td>
  </tr>
  <tr>
    <td>How many cats are in the image?</td>
    <td>2</td>
    <td>Visual Question Answering</td>
  </tr>
</table>
```

└── train.py
```

## Architecture

```bash
.
├── data
│   └── ...
├── models      # 👈 You are here
│   └── ...
└── train.py     
```

We model nanoVLM after two well known and widely used architectures. Our vision backbone
(`models/vision_transformer.py`) is the standard vision transformer, more specifically Google’s
[SigLIP](https://huggingface.co/docs/transformers/en/model_doc/siglip) vision encoder. Our language
backbone follows the [Llama 3](https://huggingface.co/docs/transformers/en/model_doc/llama3) architecture.

The vision and text modalities are *aligned* using a Modality Projection module. This module takes the
image embeddings produced by the vision backbone as input, and transforms them into embeddings
compatible with the text embeddings from the embedding layer of the language model. These embeddings
are then concatenated and fed into the language decoder. The Modality Projection module consists of a
pixel shuffle operation followed by a linear layer.

| ![diagram of the model architecture](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/nanovlm/architecture.png) |
| :--: |
| The architecture of the model (Source: Authors) |

> [Pixel shuffle](https://arxiv.org/abs/1609.05158) reduces the number of image tokens, which helps

Reviewer 3: ```suggestion
[Pixel shuffle](https://huggingface.co/papers/1609.05158) reduces the number of image tokens, which helps
```

```

This script is your one-stop shop for the entire training pipeline, including:

* Dataset loading and preprocessing  
* Model initialization  
* Optimization and logging

**Configuration**

Before anything else, the script loads two configuration classes from `models/config.py`:

* `TrainConfig`: Configuration parameters useful for training, like learning rates, checkpoint paths, etc.  
* `VLMConfig`: The configuration parameters used to initialize the VLM, like hidden dimensions, number of attention heads, etc.

**Data Loading**

At the heart of the data pipeline is the `get_dataloaders` function. It:

* Loads datasets via Hugging Face’s `load_dataset` API.  
* Combines and shuffles multiple datasets (if provided).  
* Applies a train/val split via indexing.  
* Wraps them in custom datasets (`VQADataset`, `MMStarDataset`) and collators (`VQACollator`, `MMStarCollator`).

> [!TIP]  
> A helpful flag here is `data_cutoff_idx`, useful for debugging on small subsets.

**Model Initialization**

The model is built via the `VisionLanguageModel` class. If you're resuming from a checkpoint, it’s as easy as:

```python
from models.vision_language_model import VisionLanguageModel

> model = VisionLanguageModel.from_pretrained(path)

Reviewer 4: ```suggestion
model = VisionLanguageModel.from_pretrained(model_path)
```

Reviewer 4: nit: ckpt load and the new model save path are same otherwise

Because the modality projector (`MP`) is freshly initialized while the backbones are pre-trained, the
optimizer is split into two parameter groups, each with its own learning rate:

* A higher LR for the MP  
* A smaller LR for the encoder/decoder stack

This balance ensures the MP learns quickly while preserving knowledge in the vision and language backbones.

**Training Loop**

This part is fairly standard but thoughtfully structured:

* Mixed precision is used with `torch.autocast` to improve performance.  
* A cosine learning rate schedule with linear warmup is implemented via `get_lr`.  
* Token throughput (tokens/sec) is logged per batch for performance monitoring.

Every 250 steps (configurable), the model is evaluated on the validation and `MMStar` test datasets. If accuracy improves, the model is checkpointed.

**Logging & Monitoring**

If `log_wandb` is enabled, training stats like `batch_loss`,  `val_loss`, `accuracy`, and `tokens_per_second`
are logged to Weights & Biases for real-time tracking.

Runs are auto-named using metadata like sample size, batch size, epoch count, learning rates, and the date,
all handled by the helper `get_run_name`.

**Push to Hub**

> The script doesn’t automatically push to the Hugging Face Hub, but since models are saved via:

Reviewer 4: I really think we should make this mandatory, previously @Vaibhavs10 commented the same

Reviewer 2: I agree with the point that the users should push to the hub automatically, the problem is more that we also use the same code quite extensively ourselves so if we make it the default to push to the hub, every experiment we run gets pushed there too, or we would have to change this for every new commit we do on our machines vs in the repo.

I thought about a compromise that we could change the colab notebook (since I think a lot of people start with this) to push automatically to the hub (since we do not use it), and then leave it as an option in the normal codebase. 

What do you think?

Author: Starting with colab is a very nice decision.

Reviewer: We can definitely change it in our configs if we don't want this to happen, it can just be the default, but have an option to deactivate it.

```python

> model.save_pretrained(path)

Reviewer 4: ```suggestion
model.save_pretrained(save_path)
```

| ![yoga](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/controlnet/yoga1.jpeg) |What is the woman doing? | Here in the middle she is performing yoga |

> [!TIP]  
> If you want to run inference on the trained model in a UI interface, [here](https://huggingface.co/spaces/ariG23498/nanovlm) is the Hugging Face Space for you to interact with the model. 

## Conclusion

In this blog post, we walked through what VLMs are, explored the architecture choices that power nanoVLM, and unpacked the training and inference workflows in detail.

By keeping the codebase lightweight and readable, nanoVLM aims to serve as both a learning tool and a foundation you can build upon. Whether you’re looking to understand how multi-modal inputs are aligned, or you want to train a VLM on your own dataset, this repository gives you a head start.

If you try it out, build on top of it, or just have questions we’d love to hear from you. Happy tinkering!

## References

1. [GitHub - huggingface/nanoVLM: The simplest, fastest repository for training/finetuning small-sized VLMs.](https://github.com/huggingface/nanoVLM)  

> 2. [A Dive into Vision-Language Models](https://huggingface.co/blog/vision_language_pretraining)  

Reviewer 3: I'd reorder the VLM blogs by release date.

3. [Vision Language Models Explained](https://huggingface.co/blog/vlms)  
4. [Vision Language Models (Better, faster, stronger)](https://huggingface.co/blog/vlms-2025)

> 5. [SmolVLM: Redefining small and efficient multimodal models](https://arxiv.org/abs/2504.05299)

Reviewer 3: ```suggestion
5. [SmolVLM: Redefining small and efficient multimodal models](https://huggingface.co/papers/2504.05299)
```

At its heart, nanoVLM is a **toolkit** that helps you build and train a model that can understand both
images and text, and then generate text based on that. The beauty of nanoVLM lies in its *simplicity*.
The entire codebase is intentionally kept *minimal* and *readable*, making it perfect for beginners or
anyone who wants to peek under the hood of VLMs without getting overwhelmed.

In this blog post, we cover the core ideas behind the project and provide a simple way to interact
with the repository. We not only go into the details of the project but also encapsulate all of it
so that you can quickly get started.

## Table of contents:

- [What is a Vision Language Model?](#what-is-a-vision-language-model?)
- [Working with the repository](#working-with-the-repository)
- [Architecture](#architecture)
- [Train your own VLM](#train-your-own-vlm)
- [Run inference on a pre-trained model](#run-inference-on-a-pre-trained-model)
- [Conclusion](#conclusion)
- [References](#references)

## TL;DR

You can start training a Vision Language Model using our nanoVLM toolkit by following these steps:

```bash
# Clone the repo

> git clone https://github.com/huggingface/transformers.git

Reviewer 5: ```suggestion
git clone https://github.com/huggingface/nanoVLM.git
```

# nanoVLM: The simplest repository to train your VLM in pure PyTorch

[**nanoVLM**](https://github.com/huggingface/nanoVLM) is the *simplest* way to get started with
**training** your very own Vision Language Model (VLM) using pure PyTorch. It is lightweight *toolkit*
which allows you to launch a VLM training on a [free tier colab notebook](https://colab.research.google.com/github/huggingface/nanoVLM/blob/main/nanoVLM.ipynb).

> We were inspired by [Andrej Karpathy](https://karpathy.ai/)’s [nanoGPT](https://github.com/karpathy/nanoGPT),

> and provide a similar project for the vision domain.

Reviewer 2: <img width="981" alt="image" src="https://github.com/user-attachments/assets/e1b1d042-18e7-48d8-a890-9ab4296982f1" />
